homeassistant/components/withings/common.py

```diff
@@ -280,58 +655,393 @@ async def update_sleep_summary(self) -> SleepGetSummaryResponse:
         )
 
-        def function():
-            return self._api.sleep_get_summary(lastupdate=yesterday_noon)
+        # Pre-populate values with None.
+        values: Dict[GetSleepSummaryField, int] = {
+            field: None for field in GetSleepSummaryField
+        }
+
+        def add_value(measurement: GetSleepSummaryField, value: int) -> None:
+            if value is None:
+                return
 
-        self._sleep_summary = await self.call(
-            function, throttle_domain="update_sleep_summary"
+            total = values[measurement]
+            if total is None:
+                total = 0
+
+            values[measurement] = total + value
+
+        # Sum the values.
+        for serie in response.series:
+            data = serie.data
+
+            add_value(GetSleepSummaryField.REM_SLEEP_DURATION, data.remsleepduration)
+            add_value(GetSleepSummaryField.WAKEUP_DURATION, data.wakeupduration)
+            add_value(
+                GetSleepSummaryField.LIGHT_SLEEP_DURATION, data.lightsleepduration
+            )
+            add_value(GetSleepSummaryField.DEEP_SLEEP_DURATION, data.deepsleepduration)
+            add_value(GetSleepSummaryField.WAKEUP_COUNT, data.wakeupcount)
+            add_value(GetSleepSummaryField.DURATION_TO_SLEEP, data.durationtosleep)
+            add_value(GetSleepSummaryField.DURATION_TO_WAKEUP, data.durationtowakeup)
+            add_value(GetSleepSummaryField.HR_AVERAGE, data.hr_average)
+            add_value(GetSleepSummaryField.HR_MIN, data.hr_min)
+            add_value(GetSleepSummaryField.HR_MAX, data.hr_max)
+            add_value(GetSleepSummaryField.RR_AVERAGE, data.rr_average)
+            add_value(GetSleepSummaryField.RR_MIN, data.rr_min)
+            add_value(GetSleepSummaryField.RR_MAX, data.rr_max)
+
+        # Set the value to the calculated total or None.
+        for field, value in values.items():
+            if value is not None:
+                value = round(value, 4)
+            await self._async_set_data_for_measure_type(field, value)
+
+    async def _async_set_data_for_measure_type(self, measure_type, value: Any) -> None:
+        attributes = [
+            attribute
+            for attribute in WITHINGS_ATTRIBUTES
+            if attribute.measute_type == measure_type
+        ]
+
+        for attribute in attributes:
+            await self.async_set_data(attribute, value)
+
+    async def async_set_data(self, attribute: WithingsAttribute, value: Any) -> None:
+        """Set data for a specific measurement."""
+        entity_id = await async_get_entity_id(
+            self._hass, attribute, self._profile, self._user_id
         )
 
-        return self._sleep_summary
+        if not entity_id:
+            return
+
+        if attribute.platform == BINARY_SENSOR_DOMAIN:
+            state = STATE_ON if value is True else STATE_OFF
+        else:
+            state = value
+
+        self._hass.states.async_set(
```

Comment:
This is not allowed. You have to use entities and platforms.

Reply:
WithingsDataManager now keeps a dict of state values and the entities now pull from that dictionary during async_update().